From: Luca Weiss <luca.weiss@fairphone.com>
Date: Fri, 4 Feb 2022 16:01:01 +0100
Subject: ASoC: aw882xx: Fix up for mainline

Among others, mainline q6afe doesn't set this aw_afe_fn
---
 sound/soc/codecs/Makefile                 |  3 +-
 sound/soc/codecs/aw882xx/aw882xx.c        | 88 ++++++++++++-------------------
 sound/soc/codecs/aw882xx/aw882xx.h        |  1 -
 sound/soc/codecs/aw882xx/awinic_cali.c    | 82 ++++++++++++++--------------
 sound/soc/codecs/aw882xx/awinic_dsp.c     | 12 ++++-
 sound/soc/codecs/aw882xx/awinic_monitor.c |  6 +--
 6 files changed, 89 insertions(+), 103 deletions(-)

diff --git a/sound/soc/codecs/Makefile b/sound/soc/codecs/Makefile
index 732f5b0..aaf8931 100644
--- a/sound/soc/codecs/Makefile
+++ b/sound/soc/codecs/Makefile
@@ -877,4 +877,5 @@ obj-$(CONFIG_SND_SOC_LPASS_TX_MACRO)	+= snd-soc-lpass-tx-macro.o
 # Mux
 obj-$(CONFIG_SND_SOC_SIMPLE_MUX)	+= snd-soc-simple-mux.o
 
-obj-$(CONFIG_SND_SMARTPA_AW882XX)	+= aw882xx/aw882xx.o aw882xx/awinic_monitor.o aw882xx/awinic_cali.o aw882xx/awinic_dsp.o
+snd-soc-aw882xx-y := aw882xx/aw882xx.o aw882xx/awinic_monitor.o aw882xx/awinic_cali.o aw882xx/awinic_dsp.o
+obj-$(CONFIG_SND_SMARTPA_AW882XX)	+= snd-soc-aw882xx.o
diff --git a/sound/soc/codecs/aw882xx/aw882xx.c b/sound/soc/codecs/aw882xx/aw882xx.c
index 633243e..d5df72f 100644
--- a/sound/soc/codecs/aw882xx/aw882xx.c
+++ b/sound/soc/codecs/aw882xx/aw882xx.c
@@ -32,6 +32,7 @@
 #include <sound/tlv.h>
 #include <sound/control.h>
 #include <linux/uaccess.h>
+#include <linux/pinctrl/consumer.h>
 #include "aw882xx.h"
 #include "aw882xx_reg.h"
 #include "awinic_cali.h"
@@ -66,7 +67,7 @@
 static DEFINE_MUTEX(g_aw_copp_lock);
 unsigned int g_copp_en = 0;
 
-extern int i2c_check_status_create(char *name,int value);
+//extern int i2c_check_status_create(char *name,int value);
 bool aw882xx_i2c_check = false;
 static int aw_select_pin_ctl(struct aw882xx *aw882xx, const char *name);
 /******************************************************
@@ -94,18 +95,14 @@ static unsigned int aw882xx_mode_cfg_shift[AW882XX_MODE_SHIFT_MAX] = {
  ******************************************************/
 #ifdef AW_KERNEL_VER_OVER_4_19_1
 static const struct aw_componet_codec_ops aw_componet_codec_ops = {
-	.aw_snd_soc_kcontrol_codec = snd_soc_kcontrol_component,
 	.aw_snd_soc_codec_get_drvdata = snd_soc_component_get_drvdata,
 	.aw_snd_soc_add_codec_controls = snd_soc_add_component_controls,
-	.aw_snd_soc_unregister_codec = snd_soc_unregister_component,
-	.aw_snd_soc_register_codec = snd_soc_register_component,
+	.aw_snd_soc_register_codec = devm_snd_soc_register_component,
 };
 #else
 static const struct aw_componet_codec_ops aw_componet_codec_ops = {
-	.aw_snd_soc_kcontrol_codec = snd_soc_kcontrol_codec,
 	.aw_snd_soc_codec_get_drvdata = snd_soc_codec_get_drvdata,
 	.aw_snd_soc_add_codec_controls = snd_soc_add_codec_controls,
-	.aw_snd_soc_unregister_codec = snd_soc_unregister_codec,
 	.aw_snd_soc_register_codec = snd_soc_register_codec,
 };
 #endif
@@ -445,7 +442,7 @@ static int aw882xx_sysst_check(struct aw882xx *aw882xx)
 	return ret;
 }
 
-int aw882xx_get_sysint(struct aw882xx *aw882xx, uint16_t *sysint)
+static int aw882xx_get_sysint(struct aw882xx *aw882xx, uint16_t *sysint)
 {
 	int ret = -1;
 	unsigned int reg_val = 0;
@@ -774,7 +771,7 @@ static int aw882xx_load_reg(struct aw882xx *aw882xx)
 {
 	aw_dev_info(aw882xx->dev, "%s: enter\n", __func__);
 
-	return request_firmware_nowait(THIS_MODULE, FW_ACTION_HOTPLUG,
+	return request_firmware_nowait(THIS_MODULE, FW_ACTION_UEVENT,
 		aw882xx->chan_info.bin_cfg_name[aw882xx->cfg_num],
 		aw882xx->dev, GFP_KERNEL,
 		aw882xx, aw882xx_reg_loaded);
@@ -911,8 +908,7 @@ static int aw882xx_volume_info(struct snd_kcontrol *kcontrol,
 static int aw882xx_volume_get(struct snd_kcontrol *kcontrol,
 	struct snd_ctl_elem_value *ucontrol)
 {
-	aw_snd_soc_codec_t *codec =
-		aw_componet_codec_ops.aw_snd_soc_kcontrol_codec(kcontrol);
+	aw_snd_soc_codec_t *codec = snd_kcontrol_chip(kcontrol);
 	struct aw882xx *aw882xx =
 		aw_componet_codec_ops.aw_snd_soc_codec_get_drvdata(codec);
 	unsigned int reg_val = 0;
@@ -931,8 +927,7 @@ static int aw882xx_volume_put(struct snd_kcontrol *kcontrol,
 {
 	struct soc_mixer_control *mc =
 		(struct soc_mixer_control *) kcontrol->private_value;
-	aw_snd_soc_codec_t *codec =
-		aw_componet_codec_ops.aw_snd_soc_kcontrol_codec(kcontrol);
+	aw_snd_soc_codec_t *codec = snd_kcontrol_chip(kcontrol);
 	struct aw882xx *aw882xx =
 		aw_componet_codec_ops.aw_snd_soc_codec_get_drvdata(codec);
 	unsigned int value = 0;
@@ -979,8 +974,7 @@ static struct snd_kcontrol_new aw882xx_volume = {
 static int aw882xx_mode_get(struct snd_kcontrol *kcontrol,
 	struct snd_ctl_elem_value *ucontrol)
 {
-	aw_snd_soc_codec_t *codec =
-		aw_componet_codec_ops.aw_snd_soc_kcontrol_codec(kcontrol);
+	aw_snd_soc_codec_t *codec = snd_kcontrol_chip(kcontrol);
 	struct aw882xx *aw882xx =
 		aw_componet_codec_ops.aw_snd_soc_codec_get_drvdata(codec);
 
@@ -995,8 +989,7 @@ static int aw882xx_mode_get(struct snd_kcontrol *kcontrol,
 static int aw882xx_mode_set(struct snd_kcontrol *kcontrol,
 	struct snd_ctl_elem_value *ucontrol)
 {
-	aw_snd_soc_codec_t *codec =
-		aw_componet_codec_ops.aw_snd_soc_kcontrol_codec(kcontrol);
+	aw_snd_soc_codec_t *codec = snd_kcontrol_chip(kcontrol);
 	struct aw882xx *aw882xx =
 		aw_componet_codec_ops.aw_snd_soc_codec_get_drvdata(codec);
 
@@ -1020,8 +1013,7 @@ static int aw882xx_rx_enable_get(struct snd_kcontrol *kcontrol,
 	int ret = -EINVAL;
 	uint32_t ctrl_value = 0;
 
-	aw_snd_soc_codec_t *codec =
-		aw_componet_codec_ops.aw_snd_soc_kcontrol_codec(kcontrol);
+	aw_snd_soc_codec_t *codec = snd_kcontrol_chip(kcontrol);
 	struct aw882xx *aw882xx =
 		aw_componet_codec_ops.aw_snd_soc_codec_get_drvdata(codec);
 
@@ -1042,8 +1034,7 @@ static int aw882xx_rx_enable_set(struct snd_kcontrol *kcontrol,
 {
 	int ret = -EINVAL;
 	uint32_t ctrl_value = 0;
-	aw_snd_soc_codec_t *codec =
-		aw_componet_codec_ops.aw_snd_soc_kcontrol_codec(kcontrol);
+	aw_snd_soc_codec_t *codec = snd_kcontrol_chip(kcontrol);
 	struct aw882xx *aw882xx =
 		aw_componet_codec_ops.aw_snd_soc_codec_get_drvdata(codec);
 
@@ -1064,8 +1055,7 @@ static int aw882xx_tx_enable_get(struct snd_kcontrol *kcontrol,
 {
 	int ret = -EINVAL;
 	uint32_t ctrl_value = 0;
-	aw_snd_soc_codec_t *codec =
-		aw_componet_codec_ops.aw_snd_soc_kcontrol_codec(kcontrol);
+	aw_snd_soc_codec_t *codec = snd_kcontrol_chip(kcontrol);
 	struct aw882xx *aw882xx =
 		aw_componet_codec_ops.aw_snd_soc_codec_get_drvdata(codec);
 
@@ -1087,8 +1077,7 @@ static int aw882xx_tx_enable_set(struct snd_kcontrol *kcontrol,
 	int ret = -EINVAL;
 	uint32_t ctrl_value = 0;
 
-	aw_snd_soc_codec_t *codec =
-		aw_componet_codec_ops.aw_snd_soc_kcontrol_codec(kcontrol);
+	aw_snd_soc_codec_t *codec = snd_kcontrol_chip(kcontrol);
 	struct aw882xx *aw882xx =
 		aw_componet_codec_ops.aw_snd_soc_codec_get_drvdata(codec);
 
@@ -1108,8 +1097,7 @@ static int aw882xx_tx_enable_set(struct snd_kcontrol *kcontrol,
 static int aw882xx_pa_switch_get(struct snd_kcontrol *kcontrol,
 		struct snd_ctl_elem_value *ucontrol)
 {
-	aw_snd_soc_codec_t *codec =
-		aw_componet_codec_ops.aw_snd_soc_kcontrol_codec(kcontrol);
+	aw_snd_soc_codec_t *codec = snd_kcontrol_chip(kcontrol);
 	struct aw882xx *aw882xx =
 		aw_componet_codec_ops.aw_snd_soc_codec_get_drvdata(codec);
 	int *pa_switch_ptr = &aw882xx->aw882xx_pa_switch;
@@ -1126,8 +1114,7 @@ static int aw882xx_pa_switch_set(struct snd_kcontrol *kcontrol,
 		struct snd_ctl_elem_value *ucontrol)
 {
 	int ret;
-	aw_snd_soc_codec_t *codec =
-		aw_componet_codec_ops.aw_snd_soc_kcontrol_codec(kcontrol);
+	aw_snd_soc_codec_t *codec = snd_kcontrol_chip(kcontrol);
 	struct aw882xx *aw882xx =
 		aw_componet_codec_ops.aw_snd_soc_codec_get_drvdata(codec);
 	int *aw882xx_pa_switch = &aw882xx->aw882xx_pa_switch;
@@ -1167,8 +1154,7 @@ static int aw882xx_copp_en_set(struct snd_kcontrol *kcontrol,
 	struct snd_ctl_elem_value *ucontrol)
 {
 	int ret = -EINVAL;
-	aw_snd_soc_codec_t *codec =
-		aw_componet_codec_ops.aw_snd_soc_kcontrol_codec(kcontrol);
+	aw_snd_soc_codec_t *codec = snd_kcontrol_chip(kcontrol);
 	struct aw882xx *aw882xx =
 		aw_componet_codec_ops.aw_snd_soc_codec_get_drvdata(codec);
 
@@ -1190,8 +1176,7 @@ static int aw882xx_copp_en_set(struct snd_kcontrol *kcontrol,
 static int aw882xx_copp_en_get(struct snd_kcontrol *kcontrol,
 	struct snd_ctl_elem_value *ucontrol)
 {
-	aw_snd_soc_codec_t *codec =
-		aw_componet_codec_ops.aw_snd_soc_kcontrol_codec(kcontrol);
+	aw_snd_soc_codec_t *codec = snd_kcontrol_chip(kcontrol);
 	struct aw882xx *aw882xx =
 		aw_componet_codec_ops.aw_snd_soc_codec_get_drvdata(codec);
 
@@ -1207,8 +1192,7 @@ static int aw882xx_copp_en_get(struct snd_kcontrol *kcontrol,
 static int aw882xx_hagc_en_get(struct snd_kcontrol *kcontrol,
 	struct snd_ctl_elem_value *ucontrol)
 {
-	aw_snd_soc_codec_t *codec =
-		aw_componet_codec_ops.aw_snd_soc_kcontrol_codec(kcontrol);
+	aw_snd_soc_codec_t *codec = snd_kcontrol_chip(kcontrol);
 	struct aw882xx *aw882xx =
 		aw_componet_codec_ops.aw_snd_soc_codec_get_drvdata(codec);
 
@@ -1219,8 +1203,7 @@ static int aw882xx_hagc_en_get(struct snd_kcontrol *kcontrol,
 static int aw882xx_hagc_en_set(struct snd_kcontrol *kcontrol,
 	struct snd_ctl_elem_value *ucontrol)
 {
-	aw_snd_soc_codec_t *codec =
-		aw_componet_codec_ops.aw_snd_soc_kcontrol_codec(kcontrol);
+	aw_snd_soc_codec_t *codec = snd_kcontrol_chip(kcontrol);
 	struct aw882xx *aw882xx =
 		aw_componet_codec_ops.aw_snd_soc_codec_get_drvdata(codec);
 
@@ -1293,7 +1276,7 @@ static void aw882xx_kcontrol_append_suffix(struct aw882xx *aw882xx,
 						dst_control, num);
 }
 
-void aw882xx_add_codec_controls(struct aw882xx *aw882xx)
+static void aw882xx_add_codec_controls(struct aw882xx *aw882xx)
 {
 	aw_dev_info(aw882xx->dev, "%s: enter\n", __func__);
 
@@ -1334,7 +1317,7 @@ static int aw882xx_set_fmt(struct snd_soc_dai *dai, unsigned int fmt)
 	switch (fmt & SND_SOC_DAIFMT_FORMAT_MASK) {
 	case SND_SOC_DAIFMT_I2S:
 		if ((fmt & SND_SOC_DAIFMT_MASTER_MASK) !=
-			SND_SOC_DAIFMT_CBS_CFS) {
+			SND_SOC_DAIFMT_CBC_CFC) {
 			aw_dev_err(codec->dev, "%s: invalid codec master mode\n",
 				__func__);
 			return -EINVAL;
@@ -1531,7 +1514,7 @@ static struct snd_soc_dai_driver aw882xx_dai[] = {
 			.formats = AW882XX_FORMATS,
 		 },
 		.ops = &aw882xx_dai_ops,
-		.symmetric_rates = 1,
+		.symmetric_rate = 1,
 	},
 };
 
@@ -1760,7 +1743,7 @@ static void aw882xx_parse_dt(struct device *dev, struct aw882xx *aw882xx,
 	aw882xx_parse_monitor_dt(&aw882xx->monitor);
 }
 
-void aw882xx_hw_reset(struct aw882xx *aw882xx)
+static void aw882xx_hw_reset(struct aw882xx *aw882xx)
 {
     aw_dev_info(aw882xx->dev, "%s: enter\n", __func__);
     if (IS_ERR(aw882xx->aw_pinctrl)) {
@@ -1988,8 +1971,7 @@ static int aw_select_pin_ctl(struct aw882xx *aw882xx, const char *name)
  * i2c driver
  *
  ******************************************************/
-static int aw882xx_i2c_probe(struct i2c_client *i2c,
-	const struct i2c_device_id *id)
+static int aw882xx_i2c_probe(struct i2c_client *i2c)
 {
     struct snd_soc_dai_driver *dai = NULL;
     struct aw882xx *aw882xx = NULL;
@@ -2030,7 +2012,7 @@ static int aw882xx_i2c_probe(struct i2c_client *i2c,
     aw882xx->aw_pinctrl = devm_pinctrl_get(aw882xx->dev);
     if (IS_ERR(aw882xx->aw_pinctrl)) {
         ret = PTR_ERR(aw882xx->aw_pinctrl);
-        aw_dev_err(aw882xx->dev,  "Cannot get pinctrl\n", ret);
+        aw_dev_err(aw882xx->dev,  "Cannot get pinctrl: %d\n", ret);
         return ret;
     }
 
@@ -2070,7 +2052,7 @@ static int aw882xx_i2c_probe(struct i2c_client *i2c,
 			return ret;
 
 		ret = devm_gpio_request_one(&i2c->dev, aw882xx->irq_gpio,
-			GPIOF_DIR_IN, aw882xx_int);
+			GPIOF_IN, aw882xx_int);
 		if (ret) {
 			aw_dev_err(&i2c->dev, "%s: int request failed\n",
 				__func__);
@@ -2173,22 +2155,21 @@ static int aw882xx_i2c_probe(struct i2c_client *i2c,
 	aw882xx->is_power_on = AW882XX_PA_CLOSE_ST;
 	aw_dev_dbg(aw882xx->dev, "%s: probe completed successfully!\n",
 		__func__);
-	if(!aw882xx_i2c_check)
-	{
-		i2c_check_status_create("audio_smpart_pa",1);
-		aw882xx_i2c_check = true;
-	}
+	//if(!aw882xx_i2c_check)
+	//{
+	//	i2c_check_status_create("audio_smpart_pa",1);
+	//	aw882xx_i2c_check = true;
+	//}
 	return 0;
 
 
 err_sysfs:
 err_irq:
-	aw_componet_codec_ops.aw_snd_soc_unregister_codec(&i2c->dev);
 
 	return ret;
 }
 
-static int aw882xx_i2c_remove(struct i2c_client *i2c)
+static void aw882xx_i2c_remove(struct i2c_client *i2c)
 {
 	struct aw882xx *aw882xx = i2c_get_clientdata(i2c);
 
@@ -2196,10 +2177,6 @@ static int aw882xx_i2c_remove(struct i2c_client *i2c)
 
 	aw_cali_deinit(&aw882xx->cali);
 	aw882xx_monitor_deinit(&aw882xx->monitor);
-
-	aw_componet_codec_ops.aw_snd_soc_unregister_codec(&i2c->dev);
-
-	return 0;
 }
 
 static const struct i2c_device_id aw882xx_i2c_id[] = {
@@ -2214,6 +2191,7 @@ static struct of_device_id aw882xx_dt_match[] = {
 	{ .compatible = "awinic,aw882xx_smartpa_r" },
 	{ },
 };
+MODULE_DEVICE_TABLE(of, aw882xx_dt_match);
 
 static struct i2c_driver aw882xx_i2c_driver = {
 	.driver = {
diff --git a/sound/soc/codecs/aw882xx/aw882xx.h b/sound/soc/codecs/aw882xx/aw882xx.h
index 60237b3..fbf589f 100644
--- a/sound/soc/codecs/aw882xx/aw882xx.h
+++ b/sound/soc/codecs/aw882xx/aw882xx.h
@@ -99,7 +99,6 @@ typedef struct snd_soc_codec_driver aw_snd_soc_codec_driver_t;
 #endif
 
 struct aw_componet_codec_ops {
-	aw_snd_soc_codec_t *(*aw_snd_soc_kcontrol_codec)(struct snd_kcontrol *kcontrol);
 	void *(*aw_snd_soc_codec_get_drvdata)(aw_snd_soc_codec_t *codec);
 	int (*aw_snd_soc_add_codec_controls)(aw_snd_soc_codec_t *codec,
 		const struct snd_kcontrol_new *controls,
diff --git a/sound/soc/codecs/aw882xx/awinic_cali.c b/sound/soc/codecs/aw882xx/awinic_cali.c
index 9eba9ac..0716aeb 100644
--- a/sound/soc/codecs/aw882xx/awinic_cali.c
+++ b/sound/soc/codecs/aw882xx/awinic_cali.c
@@ -38,53 +38,53 @@ static DEFINE_MUTEX(g_aw_dsp_msg_lock);
 #define AW_INT_DEC_DIGIT 10
 static int aw882xx_write_cali_re_to_file(int32_t cali_re, int channel)
 {
-	struct file *fp = NULL;
+	//struct file *fp = NULL;
 	char buf[50] = {0};
 	loff_t pos = 0;
-	mm_segment_t fs;
-
-	fp = filp_open(AWINIC_CALI_FILE, O_RDWR | O_CREAT, 0664);
-	if (IS_ERR(fp)) {
-		pr_err("%s:channel:%d open %s failed!\n",
-			__func__, channel, AWINIC_CALI_FILE);
-		return -EINVAL;
-	}
+	//mm_segment_t fs;
+
+	//fp = filp_open(AWINIC_CALI_FILE, O_RDWR | O_CREAT, 0664);
+	//if (IS_ERR(fp)) {
+	//	pr_err("%s:channel:%d open %s failed!\n",
+	//		__func__, channel, AWINIC_CALI_FILE);
+	//	return -EINVAL;
+	//}
 	if (channel == AW882XX_CHANNLE_RIGHT)
 		pos = AW_INT_DEC_DIGIT;
 
 	cali_re = FIXED_RE_TO_MOHM(cali_re);
 	snprintf(buf, sizeof(buf), "%10d", cali_re);
 
-	fs = get_fs();
-	set_fs(KERNEL_DS);
+	//fs = get_fs();
+	//set_fs(KERNEL_DS);
 
-	vfs_write(fp, buf, strlen(buf), &pos);
+	//vfs_write(fp, buf, strlen(buf), &pos);
 
-	set_fs(fs);
+	//set_fs(fs);
 
 	pr_info("%s: channel:%d buf:%s cali_re:%d\n",
 		__func__, channel, buf, cali_re);
 
-	filp_close(fp, NULL);
+	//filp_close(fp, NULL);
 	return 0;
 }
 
 static int aw882xx_get_cali_re_from_file(int32_t *cali_re, int channel)
 {
-	struct file *fp = NULL;
+	//struct file *fp = NULL;
 	/*struct inode *node;*/
 	int f_size;
 	char *buf = NULL;
 	int32_t int_cali_re = 0;
 	loff_t pos = 0;
-	mm_segment_t fs;
+	//mm_segment_t fs;
 
-	fp = filp_open(AWINIC_CALI_FILE, O_RDONLY, 0);
-	if (IS_ERR(fp)) {
-		pr_err("%s:channel:%d open %s failed!\n",
-			__func__, channel, AWINIC_CALI_FILE);
-		return -EINVAL;
-	}
+	//fp = filp_open(AWINIC_CALI_FILE, O_RDONLY, 0);
+	//if (IS_ERR(fp)) {
+	//	pr_err("%s:channel:%d open %s failed!\n",
+	//		__func__, channel, AWINIC_CALI_FILE);
+	//	return -EINVAL;
+	//}
 
 	if (channel == AW882XX_CHANNLE_RIGHT)
 		pos = AW_INT_DEC_DIGIT;
@@ -97,16 +97,16 @@ static int aw882xx_get_cali_re_from_file(int32_t *cali_re, int channel)
 	if (!buf) {
 		pr_err("%s: channel:%d malloc mem %d failed!\n",
 			__func__, channel, f_size);
-		filp_close(fp, NULL);
+		//filp_close(fp, NULL);
 		return -EINVAL;
 	}
 
-	fs = get_fs();
-	set_fs(KERNEL_DS);
+	//fs = get_fs();
+	//set_fs(KERNEL_DS);
 
-	vfs_read(fp, buf, f_size, &pos);
+	//vfs_read(fp, buf, f_size, &pos);
 
-	set_fs(fs);
+	//set_fs(fs);
 
 	if (sscanf(buf, "%d", &int_cali_re) == 1)
 		*cali_re = MOHM_TO_FIXED_RE(int_cali_re);
@@ -118,7 +118,7 @@ static int aw882xx_get_cali_re_from_file(int32_t *cali_re, int channel)
 
 	kfree(buf);
 	buf = NULL;
-	filp_close(fp, NULL);
+	//filp_close(fp, NULL);
 
 	return  0;
 
@@ -126,7 +126,7 @@ static int aw882xx_get_cali_re_from_file(int32_t *cali_re, int channel)
 #endif
 
  /*custom need add to set/get cali_re form/to nv*/
-int aw882xx_set_cali_re_to_nvram(int32_t cali_re, int32_t channel)
+static int aw882xx_set_cali_re_to_nvram(int32_t cali_re, int32_t channel)
 {
 	/*custom add, if success return value is 0, else -1*/
 #ifdef AW_CALI_STORE_EXAMPLE
@@ -135,7 +135,7 @@ int aw882xx_set_cali_re_to_nvram(int32_t cali_re, int32_t channel)
 	return -EBUSY;
 #endif
 }
-int aw882xx_get_cali_re_from_nvram(int32_t *cali_re, int32_t channel)
+static int aw882xx_get_cali_re_from_nvram(int32_t *cali_re, int32_t channel)
 {
 	/*custom add, if success return value is 0 , else -1*/
 #ifdef AW_CALI_STORE_EXAMPLE
@@ -474,7 +474,7 @@ static int aw882xx_cali_start_up(struct aw882xx *aw882xx)
 static int R0_MAX = 15000;
 static int R0_MIN = 5000;
 
-int  aw_cali_range_open(struct inode *inode, struct file *file)
+static int  aw_cali_range_open(struct inode *inode, struct file *file)
 {
 	struct aw882xx *aw882xx = (struct aw882xx *)inode->i_private;
 
@@ -483,7 +483,7 @@ int  aw_cali_range_open(struct inode *inode, struct file *file)
 	return 0;
 }
 
-ssize_t aw_cali_range_read(struct file *file,
+static ssize_t aw_cali_range_read(struct file *file,
 	char __user *buf, size_t len, loff_t *ppos)
 {
 	int ret;
@@ -510,7 +510,7 @@ ssize_t aw_cali_range_read(struct file *file,
 	return ret;
 }
 
-ssize_t aw_cali_range_write(struct file *file,
+static ssize_t aw_cali_range_write(struct file *file,
 	const char __user *buf, size_t len, loff_t *ppos)
 {
 	struct aw882xx *aw882xx = (struct aw882xx *)file->private_data;
@@ -541,7 +541,7 @@ static const struct file_operations aw_cali_range_fops = {
 	.write = aw_cali_range_write,
 };
 
-int  aw_cali_open(struct inode *inode, struct file *file)
+static int  aw_cali_open(struct inode *inode, struct file *file)
 {
 	struct aw882xx *aw882xx = (struct aw882xx *)inode->i_private;
 
@@ -550,7 +550,7 @@ int  aw_cali_open(struct inode *inode, struct file *file)
 	return 0;
 }
 
-ssize_t aw_cali_read(struct file *file,
+static ssize_t aw_cali_read(struct file *file,
 	char __user *buf, size_t len, loff_t *ppos)
 {
 	int ret;
@@ -585,7 +585,7 @@ ssize_t aw_cali_read(struct file *file,
 	return simple_read_from_buffer(buf, len, ppos, ret_value, ret);
 }
 
-ssize_t aw_cali_write(struct file *file,
+static ssize_t aw_cali_write(struct file *file,
 	const char __user *buf, size_t len, loff_t *ppos)
 {
 
@@ -598,7 +598,7 @@ static const struct file_operations aw_cali_fops = {
 	.write = aw_cali_write,
 };
 
-int  aw_f0_open(struct inode *inode, struct file *file)
+static int  aw_f0_open(struct inode *inode, struct file *file)
 {
 	struct aw882xx *aw882xx = (struct aw882xx *)inode->i_private;
 
@@ -607,7 +607,7 @@ int  aw_f0_open(struct inode *inode, struct file *file)
 	return 0;
 }
 
-ssize_t aw_f0_read(struct file *file,
+static ssize_t aw_f0_read(struct file *file,
 	char __user *buf, size_t len, loff_t *ppos)
 {
 	int ret;
@@ -651,7 +651,7 @@ static const struct file_operations aw_f0_fops = {
 	.open = aw_f0_open,
 	.read = aw_f0_read,
 };
-int  aw_cali_status_open(struct inode *inode, struct file *file)
+static int  aw_cali_status_open(struct inode *inode, struct file *file)
 {
 	struct aw882xx *aw882xx = (struct aw882xx *)inode->i_private;
 
@@ -660,7 +660,7 @@ int  aw_cali_status_open(struct inode *inode, struct file *file)
 	return 0;
 }
 
-ssize_t aw_cali_status_read(struct file *file,
+static ssize_t aw_cali_status_read(struct file *file,
 	char __user *buf, size_t len, loff_t *ppos)
 {
 	int ret;
@@ -765,7 +765,7 @@ static void aw_cali_debugfs_init(struct aw882xx *aw882xx)
 	}
 }
 
-void aw_cali_debugfs_deinit(struct aw882xx *aw882xx)
+static void aw_cali_debugfs_deinit(struct aw882xx *aw882xx)
 {
 	struct aw_dbg_cali *dbg_fs = &aw882xx->cali.dbg_fs;
 
diff --git a/sound/soc/codecs/aw882xx/awinic_dsp.c b/sound/soc/codecs/aw882xx/awinic_dsp.c
index d596c8a..1c50ad5 100644
--- a/sound/soc/codecs/aw882xx/awinic_dsp.c
+++ b/sound/soc/codecs/aw882xx/awinic_dsp.c
@@ -75,19 +75,27 @@ extern int aw_send_afe_tx_module_enable(void *buf, int size);
 #else
 static int afe_get_topology(int port_id)
 {
+	if (!aw_afe_fn)
+		return -ENODEV;
 	return aw_afe_fn->afe_get_topology(port_id);
 }
 static int aw_send_afe_cal_apr(uint32_t param_id,
 	void *buf, int cmd_size, bool write)
 {
+	if (!aw_afe_fn)
+		return -ENODEV;
 	return aw_afe_fn->aw_send_afe_cal_apr(param_id,buf,cmd_size,write);
 }
 static int aw_send_afe_rx_module_enable(void *buf, int size)
 {
+	if (!aw_afe_fn)
+		return -ENODEV;
 	return aw_afe_fn->aw_send_afe_rx_module_enable(buf,size);
 }
 static int aw_send_afe_tx_module_enable(void *buf, int size)
 {
+	if (!aw_afe_fn)
+		return -ENODEV;
 	return aw_afe_fn->aw_send_afe_tx_module_enable(buf,size);
 }
 #endif
@@ -329,7 +337,7 @@ static int aw_qcom_write_data_to_dsp(int index, void *data, int data_size, int c
 	if (ret < 0)
 		return ret;
 
-	while (try < AW_DSP_TRY_TIME) {
+	while (false /*try < AW_DSP_TRY_TIME*/) {
 		if (aw_check_dsp_ready()) {
 			mutex_lock(&g_dsp_lock);
 			ret = aw_send_afe_cal_apr(param_id, data,
@@ -356,7 +364,7 @@ static int aw_qcom_read_data_from_dsp(int index, void *data, int data_size, int
 	if (ret < 0)
 		return ret;
 
-	while (try < AW_DSP_TRY_TIME) {
+	while (false /*try < AW_DSP_TRY_TIME*/) {
 		if (aw_check_dsp_ready()) {
 			mutex_lock(&g_dsp_lock);
 			ret = aw_send_afe_cal_apr(param_id, data,
diff --git a/sound/soc/codecs/aw882xx/awinic_monitor.c b/sound/soc/codecs/aw882xx/awinic_monitor.c
index 36e16c2..00d81e2 100644
--- a/sound/soc/codecs/aw882xx/awinic_monitor.c
+++ b/sound/soc/codecs/aw882xx/awinic_monitor.c
@@ -772,7 +772,7 @@ static void aw_monitor_profile_loaded(const struct firmware *cont,
 	mutex_unlock(&g_aw_monitor_lock);
 }
 
-int aw882xx_load_monitor_profile(struct aw882xx_monitor *monitor)
+static int aw882xx_load_monitor_profile(struct aw882xx_monitor *monitor)
 {
 	int ret;
 	struct aw882xx *aw882xx = container_of(monitor,
@@ -783,7 +783,7 @@ int aw882xx_load_monitor_profile(struct aw882xx_monitor *monitor)
 			__func__, monitor->is_enable);
 		ret = 0;
 	} else {
-		ret = request_firmware_nowait(THIS_MODULE, FW_ACTION_HOTPLUG,
+		ret = request_firmware_nowait(THIS_MODULE, FW_ACTION_UEVENT,
 				AW_MONITOR_FILE,
 				aw882xx->dev, GFP_KERNEL, aw882xx,
 				aw_monitor_profile_loaded);
@@ -800,7 +800,7 @@ static void aw882xx_monitor_load_fw_work_func(struct work_struct *work)
 	aw882xx_load_monitor_profile(monitor);
 }
 
-void aw882xx_deinit_monitor_profile(struct aw882xx_monitor *monitor)
+static void aw882xx_deinit_monitor_profile(struct aw882xx_monitor *monitor)
 {
 	struct aw_monitor_cfg *monitor_cfg = monitor->monitor_cfg;
 
